var/spack/repos/builtin/packages/mfem/package.py

```diff
@@ -697,6 +699,29 @@ def install(self, spec, prefix):
         if install_em:
             install_tree('data', join_path(prefix_share, 'data'))
 
+    examples_src_dir = 'examples'
+    examples_data_dir = 'data'
+
+    @run_after('install')
+    def cache_test_sources(self):
+        """Copy the example source files after the package is installed to an
+        install test subdirectory for use during `spack test run`."""
+        self.cache_extra_test_sources(self.examples_src_dir)
+        self.cache_extra_test_sources(self.examples_data_dir)
+
+    def test(self):
+        test_dir = os.path.join(self.install_test_root, self.examples_src_dir)
+        with working_dir(test_dir, create=False):
+            test_exe = 'ex10p' if ('+mpi' in self.spec) else 'ex10'
+            make('CONFIG_MK={0}/share/mfem/config.mk'.format(self.prefix),
+                 test_exe, parallel=False)
+            self.run_test('./{0}'.format(test_exe),
```

Comment:
Am curious but how does `self.run_test(...)` know if the test passed or failed? Is it examining the return code? If non-zero does it except out from here?

Does it make sense to make some of the key arguments to the test explicit as in the example command-lines described [here](https://github.com/mfem/mfem/blob/master/examples/ex10p.cpp)? If for some reason MFEM developers decide to change default behavior of `ex10[p].cpp` (perhaps unlikely but not implausible), that would mean this test would change maybe in a way that wasn't obvious in which case, that empty `[]` arg to `self.run_test(...)` might go like `[` `-s`,`3`,`-rs`,`2`,`-dt`,`3` `]`

Also, not sure how relevant this is but this test winds up testing only MFEM base level functionality (which is appropriate for a smoke check). But, should it have the option of testing if MFEM's dependencies (e.g. SUNDIALS, PETSc, etc.) are also working by invoking a test that winds up exercising those dependencies? Maybe that is an exercise for later.

Reply:
> Am curious but how does `self.run_test(...)` know if the test passed or failed? Is it examining the return code? If non-zero does it except out from here?

It can check expected return codes (using the `status` argument) and or the expected output (using the `expected` argument).  The former can be a list of acceptable integer values.  The latter can be a list of strings to be regex matched.  See https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-install-tests.

> 
> Does it make sense to make some of the key arguments to the test explicit as in the example command-lines described [here](https://github.com/mfem/mfem/blob/master/examples/ex10p.cpp)? If for some reason MFEM developers decide to change default behavior of `ex10[p].cpp` (perhaps unlikely but not implausible), that would mean this test would change maybe in a way that wasn't obvious in which case, that empty `[]` arg to `self.run_test(...)` might go like `[` `-s`,`3`,`-rs`,`2`,`-dt`,`3` `]`
> 
> Also, not sure how relevant this is but this test winds up testing only MFEM base level functionality (which is appropriate for a smoke check). But, should it have the option of testing if MFEM's dependencies (e.g. SUNDIALS, PETSc, etc.) are also working by invoking a test that winds up exercising those dependencies? Maybe that is an exercise for later.

That's up to the mfem team.  This is intended as a preliminary smoke test to:

1) leverage the current E4S test suite for the software; and
2) demonstrate to package maintainers how to start a Spack smoke test.
2) .

Reply:
See https://github.com/E4S-Project/testsuite/blob/master/validation_tests/mfem/run.sh for the E4S validation test defined by their `run.sh`.

Reply:
> It can check expected return codes (using the `status` argument) and or the expected output (using the `expected` argument).

Ok, read that ref. Maybe not necessary to eduate little ole me here in this PR but that ref indicates no default behavior and the call instance here specifies neither `status=[]` nor `expected=[]`. So, whats gonna happen? Though it doesn't appear to be documented, am guessing default behavior for `self.run_test(...)` is as if `status=[0,]` is specified. True?

Reply:
> See https://github.com/E4S-Project/testsuite/blob/master/validation_tests/mfem/run.sh for the E4S validation test defined by their `run.sh`.

Not sure if it matters but that ref runs `ex10p` whereas this smoke check runs either `ex10` or `ex10p`. Maybe parallel build is expected default and this won't matter.

Reply:
> This is intended as a preliminary smoke test to:
> 
> 1. leverage the current E4S test suite for the software; and
> 2. demonstrate to package maintainers how to start a Spack smoke test.

BTW...not trying to be a PITA but this info would have been useful in the original PR comment to help me put on the right thinking cap :wink:

Reply:
> > It can check expected return codes (using the `status` argument) and or the expected output (using the `expected` argument).
> 
> Ok, read that ref. Maybe not necessary to eduate little ole me here in this PR but that ref indicates no default behavior and the call instance here specifies neither `status=[]` nor `expected=[]`. So, whats gonna happen? Though it doesn't appear to be documented, am guessing default behavior for `self.run_test(...)` is as if `status=[0,]` is specified. True?

You are correct.  The default expected is `[]` (so ignore output) and `status` is `0` (or no error code).  The latter is converted automatically by Spack to a list of `[0]` for checking purposes.

Reply:
> > This is intended as a preliminary smoke test to:
> > 
> > 1. leverage the current E4S test suite for the software; and
> > 2. demonstrate to package maintainers how to start a Spack smoke test.
> 
> BTW...not trying to be a PITA but this info would have been useful in the original PR comment to help me put on the right thinking cap 😉

Fair enough.  I debated updating the PR description since it's not my PR.

Reply:
> You are correct. The default expected is `[]` (so ignore output) and `status` is `0` (or no error code). The latter is converted automatically by Spack to a list of `[0]` for checking purposes.

I submitted PR #21309